language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -107,31 +111,75 @@ genModule curPkgId mod
 
-genDataDef :: PackageId -> ModuleName -> NM.NameMap Template -> DefDataType -> (([T.Text], [T.Text]), Set.Set ModuleRef)
-genDataDef curPkgId curModName tpls def = case unTypeConName (dataTypeCon def) of
+genDataDef :: PackageId -> Module -> NM.NameMap Template -> DefDataType -> (([T.Text], [T.Text]), Set.Set ModuleRef)
+genDataDef curPkgId mod tpls def = case unTypeConName (dataTypeCon def) of
     [] -> error "IMPOSSIBLE: empty type constructor name"
     _: _: _: _ -> error "IMPOSSIBLE: multi-part type constructor of more than two names"
-    [conName] -> genDefDataType curPkgId conName curModName tpls def
-    [c1, c2] -> ((makeNamespace $ map ("  " <>) typs, makeNamespace $ map ("  " <>) sers), refs)
+
+    [conName] -> genDefDataType curPkgId conName mod tpls def
+    [c1, c2] -> ((makeNamespace $ map ("  " <>) typs, []), refs)
       where
-        ((typs, sers), refs) = genDefDataType curPkgId c2 curModName tpls def
-        ns = c1 <> "_NS"
+        ((typs, _), refs) = genDefDataType curPkgId c2 mod tpls def
+        ns = c1
         makeNamespace stuff =
           [ "// eslint-disable-next-line @typescript-eslint/no-namespace"
-          , "export namespace " <> ns <> " {"] ++ stuff ++ ["} //namespace " <> ns] ++ [""]
+          , "export namespace " <> ns <> " {"] ++ stuff ++ ["} //namespace " <> ns]
 
-genDefDataType :: PackageId -> T.Text -> ModuleName -> NM.NameMap Template -> DefDataType -> (([T.Text], [T.Text]), Set.Set ModuleRef)
-genDefDataType curPkgId conName curModName tpls def =
+genDefDataType :: PackageId -> T.Text -> Module -> NM.NameMap Template -> DefDataType -> (([T.Text], [T.Text]), Set.Set ModuleRef)
+genDefDataType curPkgId conName mod tpls def =
     case dataCons def of
         DataVariant bs ->
           let
             (typs, sers) = unzip $ map genBranch bs
-            typeDesc = [""] ++ typs
-            serDesc  = ["() => jtv.oneOf<" <> conName <> typeParams <> ">("] ++ sers ++ [")"]
-          in
-          ((makeType typeDesc, makeSer serDesc), Set.unions $ map (Set.setOf typeModuleRef . snd) bs)
+            typeDesc = makeType ([""] ++ typs)
+            typ = conName <> typeParams -- Type of the variant.
+            assocDefs = assocDefDataTypes mod conName -- Associated types.
+            serDesc =
+              if not $ null paramNames -- Polymorphic type.
+              then -- Companion function.
+                let
+                  -- Any associated serializers.
+                  assocSers = map
+                    (\(n, d) -> serFromDef id n curPkgId mod tpls d)
+                    assocDefs
+                  -- The variant deserializer.
+                  function = (makeSer ( ["() => jtv.oneOf<" <> typ <> ">("] ++ sers ++ [")"]) & onLast (<> ";"));
+                  props = -- Fix the first and last line of each serializer.
+                    concatMap (onHead (erase "export const ") . onLast (<> ";")) assocSers
```

Comment:
Do we really want to erase all occurrences of `export const` or only strip a prefix? Maybe [`Data.Text.stripPrefix`](https://www.stackage.org/haddock/lts-14.6/text-1.2.3.1/Data-Text.html#v:stripPrefix) is a better fit?

Reply:
I'm also wondering if `genDefDataType` is returning the wrong thing since we have to do so much mangling on it's result. Maybe we should do some refactoring here.

Reply:
(1) Probably doesn't matter but `Data.Text.stripPrefix` better captures my intents.
(2) Maybe.